### ISSUE TITLE
DES Encryption and Decryption not working

### ISSUE THREAD

**Issue by jaytrixz** [iosdevzone/IDZSwiftCommonCrypto#9]
I'm trying to use the DES encryption using ECB mode but it's not working. I validate it using the online encryption tool that works. I changed my keys and plain text to not display our own key and text and used a different pair. IDZSwiftCommonCrypto yields different results everytime I tap my Start button. Here's my code:

```swift
import UIKit
import IDZSwiftCommonCrypto

class ViewController: UIViewController {
    
    @IBOutlet weak var keyTextField: UITextField! // key is thekey
    @IBOutlet weak var plainTextField: UITextField! // plain text is username123
    @IBOutlet weak var outputTextView: UITextView! // output should be b742acfaa07e3d05cf2dc9aaa0258fc2
    @IBOutlet weak var encryptionSegmentedControl: UISegmentedControl!
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        UIApplication.sharedApplication().statusBarStyle = .LightContent
    }

    override func didReceiveMemoryWarning() {
        super.didReceiveMemoryWarning()
        // Dispose of any resources that can be recreated.
    }
    
    
    @IBAction func startButtonTapped(sender: UIButton) {
        if encryptionSegmentedControl.selectedSegmentIndex == 0 {
            let hexKey = arrayFromHexString(convertToHexFromString(keyTextField.text!))
            let encryptedBytesArray = Cryptor(operation:.Encrypt, algorithm:.DES, options:.ECBMode, key:hexKey, iv:[UInt8]()).update(arrayFromHexString(convertToHexFromString(plainTextField.text!)))?.final()
            
            outputTextView.text = hexStringFromArray(encryptedBytesArray!)
        } else {
            let hexKey = arrayFromHexString(convertToHexFromString(keyTextField.text!))
            let decryptedBytesArray = Cryptor(operation:.Decrypt, algorithm:.DES, options:.ECBMode, key:hexKey, iv:[UInt8]()).update(plainTextField.text!)?.final()
            
            outputTextView.text = hexStringFromArray(decryptedBytesArray!)
        }
    }
    
    // MARK: - Custom
    
    @IBAction func dismissKeyboard() {
        self.view.endEditing(true)
    }
    
    func convertToHexFromString(rawString: String) -> String! {
        var hexString = "\(rawString.dataUsingEncoding(NSUTF8StringEncoding)!)"
        hexString = hexString.stringByReplacingOccurrencesOfString(" ", withString: "").stringByTrimmingCharactersInSet(NSCharacterSet.symbolCharacterSet())
        
        print("Formatted hex string: \(hexString)")
        
        return hexString
    }
    
    func encryptUsingText(hexText: String, hexKey: [UInt8]) -> String! {
        let cryptor = Cryptor(operation:.Encrypt, algorithm:.DES, options:.ECBMode, key:hexKey, iv:[UInt8]())
        let cipherText = cryptor.update(hexText)?.final()
        
        print("Bytes array: \(cipherText!)")
        
        return String(bytes: cipherText!, encoding: NSUTF8StringEncoding)
    }
    
    func decryptUsingText(rawText: String, hexKey: [UInt8]) -> String! {
        let cryptor = Cryptor(operation:.Decrypt, algorithm:.DES, options:.None, key:hexKey, iv:[UInt8]())
        let decryptedCipherText = cryptor.update(rawText)?.final()
        
        print("Bytes array: \(decryptedCipherText!)")
        
        return String(bytes: decryptedCipherText!, encoding: NSUTF8StringEncoding)
    }
    
}
```

**Comment by iosdevzone** [comment]
Jay, thanks for the report. There does seem to be something strange happening here. I'll take a look at it and get back to you.

**Comment by jaytrixz** [comment]
Thanks a lot!

**Comment by iosdevzone** [comment]
OK, so it turns out there were a few separate things going on here. Some were issues with your code, others were things that my library could have done a better job of protecting you against.

All the current ciphers supported by the library at this time are block ciphers. That means that you must either manually pad out your buffers to be an integral number of blocks or you must use the `PKCS7Padding` option. Which option you choose depends on whatever protocol you are trying to match. With the sample key, plaintext and cipher text you supplied it is plain that the site you are using is zero padding the buffers and is not using PKCS7 padding.

If everything else was working, the following code snippet *should* have got the correct answer: 
```Swift 
let blockSize = Cryptor.Algorithm.DES.blockSize()
let key = "thekey"
let plainText = zeroPad("username123", blockSize) // see commit 53b11ec for zeroPad
let expectedCipherText = arrayFromHexString("b742acfaa07e3d05cf2dc9aaa0258fc2")
let cryptor = Cryptor(operation: .Encrypt, algorithm: .DES, options: [.ECBMode], key: key, iv: [UInt8]())
let cipherText = cryptor.update(plainText)?.final()
XCTAssertEqual(expectedCipherText, cipherText!)
```

This code does not get the correct answer because the underlying `CommonCrypto` call `CCCryptorCreate`, if the key is too short, reads initialized memory and uses it as part of the key material. Whether Apple consider this a bug remains to been seen.

I've just checked in a fix to my code that detects when the key is too short and zero pads it to the shortest valid length before calling  `CCCryptorCreate`.

Before I bump the release number for CocoaPods and Carthage I want to do a little more testing, but for your purposes it should solve the problem. 

Check out the test the reference Issue9 and you can see the answers you expect being generated.

I will be traveling for the next few weeks so there could be some delay getting all this tested and released. 

Let me know if you need more help.

**Comment by jaytrixz** [comment]
Thanks a lot! I'll check and do more tests to see if everything works now. Just to add, I remember using another library that uses CommonCrypto and it also produces the same results, which may mean that there's a bug in Apple's code.

**Comment by jaytrixz** [comment]
I'm this error message in Cryptor operation `.Encrypt` saying:

`Type of expression is ambiguous without more context`

**Comment by iosdevzone** [comment]
Sorry, there was a typo in the code in my comment above. Have corrected it now. Problem was that type of initialization vector must match. It was
```swift
let cryptor = Cryptor(operation: .Encrypt, algorithm: .DES, options: [.ECBMode], key: key, iv:[UInt8]())
```
Should be 
```swift
let cryptor = Cryptor(operation: .Encrypt, algorithm: .DES, options: [.ECBMode], key: key, iv:"")
```

**Comment by jaytrixz** [comment]
It works now! Thanks a lot as always! I did tried using our company's key and text and it shows a fatal error pointing to invalid key size. How can I prevent this error from happening?

**Comment by iosdevzone** [comment]
The latest version will zero pad the key to try to make it valid, if it is shorter than the expected length... if it is too long it will error. 

Perhaps send me a code snippet demonstrating the problem. You can replace your key with "xxxx" just make sure the number of 'x's is the same as the number of characters in your company's key.

**Comment by jaytrixz** [comment]
My code is pretty much the same. I just edited my `startButtonTapped` method to this:

```swift
@IBAction func startButtonTapped(sender: UIButton) {
        if encryptionSegmentedControl.selectedSegmentIndex == 0 {
            let blockSize = Cryptor.Algorithm.DES.blockSize()
            let plainText = zeroPad(plainTextField.text!, blockSize)
            
            let encryptedBytesArray = Cryptor(operation: .Encrypt, algorithm: .DES, options: [.ECBMode], key: keyTextField.text!, iv: "").update(plainText)?.final()
            
            outputTextView.text = hexStringFromArray(encryptedBytesArray!)
        } else {
            let hexKey = arrayFromHexString(convertToHexFromString(keyTextField.text!))
            let decryptedBytesArray = Cryptor(operation:.Decrypt, algorithm:.DES, options:.ECBMode, key:hexKey, iv:[UInt8]()).update(plainTextField.text!)?.final()
            
            outputTextView.text = hexStringFromArray(decryptedBytesArray!)
        }
    }
```

**Comment by iosdevzone** [comment]
So, it the length of keyTextField is greater than 8 characters after conversion to UTF8 you will get an error. You need to figure out how your protocol handles this. It may truncate the key material or blockwise-or it or some other thing. 
Either way the key you supply must be 8 bytes long (or less and it will be zero padded)

**Comment by jaytrixz** [comment]
I tried using [this] (https://www.tools4noobs.com/online_tools/encrypt/) online encrypt tool and it's not throwing an invalid key size error. Is there any way to work around this? I cannot truncate my key unfortunately

**Comment by jaytrixz** [comment]
My colleague made it to work! Thanks a lot! :) I'll post the complete code here later.

**Comment by iosdevzone** [comment]
I believe this is solved in release 0.6.8.
